mitmproxy/proxy/layers/dns.py

```diff
@@ -40,40 +39,37 @@ class DNSLayer(layer.Layer):
     Layer that handles resolving DNS queries.
     """
 
-    flow: dns.DNSFlow
+    flows: dict[int, dns.DNSFlow]
```

Comment:
How/why do we end up having multiple DNS flows here now?

Reply:
So the history here is: We had that before, in the DNS draft. Then we moved the ID into the connection_id of proxyserver (i.e. essentially the `dict`). However, if we want to do protocol detection and fancy transparent stuff, DNS might not be at the top layer, so the layer itself will have to do it.

It's the same with QUIC. We've talked about connection migration (which currently also uses connection_id in proxyserver) and the change of `SendData` and `DataReceived` to include the remote address, that ideally should not happen. So there we will also need a layer that does the routing instead, keeping the old `SendData` and not using custom `connection_id`.

And that's actually a good thing: Having uniform connection id's is the only way to reliably do `inject_udp` in the future.